sdk/dotnet/Pulumi/Deployment/Deployment_Run.cs

```diff
@@ -22,6 +22,43 @@ public static Task<int> RunAsync(Action action)
                 return ImmutableDictionary<string, object?>.Empty;
             });
 
+        /// <summary>
+        /// <see cref="RunAsync(Action{DeploymentOutputs})"/> is an
+        /// entry-point to a Pulumi application. .NET applications should perform all startup logic
+        /// they need at the top of the program and then call it:
+        /// <para>
+        /// <c>
+        /// using System;
+        /// using Pulumi;
+        ///
+        /// async Deployment.RunAsync(outputs =>
+        /// {
+        ///     var bucket = new Bucket(args);
+        ///     outputs.Export("BucketName", bucket.BucketName);
+        /// });
+        /// </c>
+        /// </para>
+        /// <para>
+        /// Importantly: Cloud resources cannot be created outside of the lambda passed to any of the
+        /// <see cref="Deployment.RunAsync(Action)"/> overloads.  Because cloud Resource construction is
+        /// inherently asynchronous, the result of this function is a <see cref="Task{T}"/> which should
+        /// then be returned or awaited.  This will ensure that any problems that are encountered during
+        /// the running of the program are properly reported.  Failure to do this may lead to the
+        /// program ending early before all resources are properly registered.
+        /// </para>
+        /// 
+        /// </summary>
+        /// <param name="contextHandler">Callback that creates stack resources.</param>
+        public static async Task<int> RunAsync(Action<DeploymentOutputs> contextHandler)
```

Comment:
What about an overload that would allow async callbacks? (so that users could await invokes, or are we consciously de-emphasizing those in favor of output-based invokes?)

Reply:
An async callback can be added when we agree on how the sync version looks like :smile: I don't know much about invokes but if these are async calls then we should definitely have an async version (also because many .NET APIs are async too)